prov/rxd/src/rxd_av.c

```diff
@@ -0,0 +1,325 @@
+}
+
+int rxd_av_insert_dg_av(struct rxd_av *av, const void *addr)
+{
+	int ret;
+	fastlock_acquire(&av->lock);
+	ret = fi_av_insert(av->dg_av, addr, 1, NULL, 0, NULL);
+	if (ret != 1)
+		goto out;
+	av->dg_av_used++;
+out:
+	fastlock_release(&av->lock);
+	return ret;
+}
+
+int rxd_dg_av_reverse_lookup(struct rxd_av *av, uint64_t start_idx,
+			      const void *addr, size_t addrlen, uint64_t *idx)
+{
+	int ret;
+	size_t i, len;
+	void *curr_addr;
+
+	len = addrlen;
+	curr_addr = calloc(1, av->addrlen);
+	if (!curr_addr)
+		return -FI_ENOMEM;
+
+	for (i = 0; i < av->dg_av_used; i++) {
+		ret = fi_av_lookup(av->dg_av, (i + start_idx) % av->dg_av_used,
+				   curr_addr, &len);
+		if (ret)
+			continue;
+		if (memcmp(curr_addr, addr, len) == 0) {
+			*idx = (i + start_idx) % av->dg_av_used;
+			goto out;
+		}
+	}
+	ret = -FI_ENODATA;
+out:
+	free(curr_addr);
+	return ret;
+}
+
+size_t rxd_av_insert_check(struct rxd_av *av, const void *addr, size_t count,
+			    fi_addr_t *fi_addr, uint64_t flags, void *context)
+{
+	size_t i, success_cnt = 0;
+	int ret, index;
+	void *curr_addr;
+	uint64_t dg_av_idx;
+
+	for (i = 0; i < count; i++) {
+		curr_addr = (char *) addr + av->addrlen * i;
+		ret = rxd_dg_av_reverse_lookup(av, i, curr_addr, av->addrlen, &dg_av_idx);
+		if (ret == -FI_ENODATA) {
+			ret = fi_av_insert(av->dg_av, curr_addr, 1, &dg_av_idx,
+					   flags, context);
+			if (ret <= 0) {
+				if (av->util_av.eq)
+					ofi_av_write_event(&av->util_av, i,
+							   (ret == 0) ? FI_EINVAL : -ret,
+							   context);
+				if (fi_addr)
+					fi_addr[i] = FI_ADDR_NOTAVAIL;
+				continue;
+			}
+		}
+
+		ret = ofi_av_insert_addr(&av->util_av, &dg_av_idx, dg_av_idx, &index);
+		if (ret) {
+			if (av->util_av.eq)
+				ofi_av_write_event(&av->util_av, i, -ret, context);
+		} else {
+			success_cnt++;
+		}
+		
+		if (fi_addr)
+			fi_addr[i] = (ret == 0) ? index : FI_ADDR_NOTAVAIL;
+	}
+	av->dg_av_used += success_cnt;
+	if (av->util_av.eq) {
+		ofi_av_write_event(&av->util_av, success_cnt, 0, context);
+		ret = 0;
+	} else {
+		ret = success_cnt;
+	}
+	return ret;
+}
+
+size_t rxd_av_insert_fast(struct rxd_av *av, const void *addr, size_t count,
+			   fi_addr_t *fi_addr, uint64_t flags, void *context)
+{
+	size_t i, num, ret, success_cnt = 0;
+	int index;
+	fi_addr_t *fi_addrs;
+
+	fi_addrs = calloc(count, sizeof(fi_addr_t));
+	if (!fi_addrs)
+		return -FI_ENOMEM;
+
+	num = fi_av_insert(av->dg_av, addr, count, fi_addrs, flags, context);
+	if (num != count) {
+		free(fi_addrs);
+		return rxd_av_insert_check(av, addr, count, fi_addr,
+					    flags, context);
+	}
+
+	for (i = 0; i < num; i++) {
+		ret = ofi_av_insert_addr(&av->util_av, &fi_addrs[i],
+					 fi_addrs[i], &index);
+		if (ret) {
+			if (av->util_av.eq)
+				ofi_av_write_event(&av->util_av, i, -ret, context);
+		} else {
+			success_cnt++;
+		}
+
+		if (fi_addr)
+			fi_addr[i] = (ret == 0) ? index : FI_ADDR_NOTAVAIL;
+	}
+
+	free(fi_addrs);
+	av->dg_av_used = success_cnt;
```

Comment:
Hmm, this path (insert_fast) will be called only when dg_av is empty. But, it makes sense to use += to avoid future errors.